crates/tpu-client/src/rpc/tpu_info.rs

```diff
@@ -0,0 +1,384 @@
+    shared: Arc<RwLock<HashMap<Pubkey, RpcTpuQuicContactInfo>>>,
+    _on_drop: Arc<OnDrop>,
+}
+
+impl RpcClusterTpuQuicInfoService {
+    ///
+    /// Get the current topology mapping of leader pubkeys to their TPU QUIC contact info.
+    ///
+    pub fn get_current_topology_mapping(&self) -> HashMap<Pubkey, RpcTpuQuicContactInfo> {
+        self.shared.read().expect("read lock").clone()
+    }
+}
+
+struct OnDrop {
+    handle: CancellationToken,
+}
+
+impl Drop for OnDrop {
+    fn drop(&mut self) {
+        self.handle.cancel();
+    }
+}
+
+///
+/// Configuration for the [`RpcClusterTpuQuicInfoService`].
+///
+#[derive(Clone)]
+pub struct RpcClusterTpuQuicInfoServiceConfig {
+    /// Interval between cluster info refreshes.
+    pub refresh_interval: std::time::Duration,
+    /// Maximum number of retry attempts for fetching cluster info.
+    pub max_retry_attempts: usize,
+}
+
+impl Default for RpcClusterTpuQuicInfoServiceConfig {
+    fn default() -> Self {
+        Self {
+            refresh_interval: DEFAULT_REFRESH_INTERVAL,
+            max_retry_attempts: 3,
+        }
+    }
+}
+
+enum Diff {
+    Upsert((Pubkey, RpcTpuQuicContactInfo)),
+    Remove(Pubkey),
+}
+
+fn detect_tpu_contact_info_diff(
+    current: &HashMap<Pubkey, RpcTpuQuicContactInfo>,
+    new: &[RpcContactInfo],
+) -> Vec<Diff> {
+    let mut to_remove = current
+        .keys()
+        .cloned()
+        .collect::<std::collections::HashSet<_>>();
+    let mut ret = new
+        .iter()
+        .filter_map(|node| {
+            let pubkey = Pubkey::from_str(&node.pubkey).expect("Pubkey::from_str");
+            let contact_info = RpcTpuQuicContactInfo::from(node);
+            to_remove.remove(&pubkey);
+            match current.get(&pubkey) {
+                Some(existing_info) => {
+                    if existing_info != &contact_info {
+                        Some(Diff::Upsert((pubkey, contact_info)))
+                    } else {
+                        None
+                    }
+                }
+                None => Some(Diff::Upsert((pubkey, contact_info))),
+            }
+        })
+        .collect::<Vec<_>>();
+
+    for removed_pubkey in to_remove {
+        ret.push(Diff::Remove(removed_pubkey));
+    }
+
+    ret
+}
+
+fn apply_diff_patch(
+    current: &mut HashMap<Pubkey, RpcTpuQuicContactInfo>,
+    diffs: impl IntoIterator<Item = Diff>,
+) {
+    for diff in diffs {
+        match diff {
+            Diff::Upsert((pubkey, contact_info)) => {
+                current.insert(pubkey, contact_info);
+            }
+            Diff::Remove(pubkey) => {
+                current.remove(&pubkey);
+            }
+        }
+    }
+}
+
+async fn cluster_info_refresh_loop(
+    rpc_client: RpcClient,
+    shared: Arc<RwLock<HashMap<Pubkey, RpcTpuQuicContactInfo>>>,
+    config: RpcClusterTpuQuicInfoServiceConfig,
+    cancellation_token: CancellationToken,
+) {
+    let mut tick = tokio::time::interval(config.refresh_interval);
```

Comment:
```suggestion
    let mut interval = tokio::time::interval(config.refresh_interval);
```

do you think calling this interval is better here? since this interval will give you the `.tick()` interface, I see the reason for the name being `tick` but it's weird reading `tick.tick()`

Reply:
good suggestion indeed